src/Compilers/CSharp/Portable/BoundTree/BoundTreeVisitors.cs

```diff
@@ -184,22 +184,23 @@ public static Location GetTooLongOrComplexExpressionErrorLocation(BoundNode node
             {
                 SyntaxNode syntax = node.Syntax;
 
-                if (!(syntax is ExpressionSyntax))
+                if (syntax is not (ExpressionSyntax or PatternSyntax))
                 {
-                    syntax = syntax.DescendantNodes(n => !(n is ExpressionSyntax)).OfType<ExpressionSyntax>().FirstOrDefault() ?? syntax;
+                    syntax = syntax.DescendantNodes(n => n is not (ExpressionSyntax or PatternSyntax)).FirstOrDefault(n => n is ExpressionSyntax or PatternSyntax) ?? syntax;
```

Comment:
>n => n is ExpressionSyntax or PatternSynta [](http://example.com/codeflow?start=118&length=42)

Why is this necessary? It looks like this condition is the opposite of the filter used for DescendantNodes.

Reply:
The predicate for `DescendentNodes` controls whether we descend into children; it stops us from returning new nodes after we hit an expression or pattern. However, it will still have given us all of the parent nodes, so we need to ignore them.